src/hex/shape.rs

```diff
@@ -51,16 +73,176 @@ impl Shape<Axial> {
         }
     }
 
+    pub fn resize(&mut self, scale: (f64, f64)) -> &Self {
+        todo!()
+    }
+
+    // Bake the transform into localspace
+    pub fn bake_transform(&mut self) -> HexShape<T> {
+        todo!()
+    }
+
     // Retrieves the hex array in local space
-    pub fn get_hexes(&self) -> Vec<Axial> {
-        self.tiles.clone()
+    pub fn get_hexes(&self) -> &[HexTile<T>] {
+        self.tiles.as_slice()
     }
 
     // Retrieves the hex array in grid space
-    pub fn get_hexes_gs(&self) -> Vec<Axial> {
+    // Because we modify the HexTile coordinates into gs we must
+    // return a cloned vector instead of a slice.
+    pub fn get_hexes_gs(&mut self) -> Vec<HexTile<T>>
+    where
+        T: Clone
+    {
         self.tiles
             .iter()
-            .map(|&coord| coord.apply_all(self.transform))
+            .map(|tile| {tile.coord.apply_all(self.transform); tile.clone()})
             .collect()
     }
 }
+
+// General shape constructors
+// Move this to another file? Inside a different module?
+pub mod shape_builders {
+    use crate::{axial, transform};
+
+    use super::*;
+
+    pub struct Inequality {
+        q_min: i32,
+        q_max: i32,
+        r_min: i32,
+        r_max: i32,
+        s_min: i32,
+        s_max: i32,
+    }
+
+    impl Inequality {
+        pub fn new(points: &[Axial]) -> Self {
+            let mut q_min = i32::MAX;
+            let mut q_max = i32::MIN;
+
+            let mut r_min = i32::MAX;
+            let mut r_max = i32::MIN;
+
+            let mut s_min = i32::MAX;
+            let mut s_max = i32::MIN;
+
+            for point in points {
+                q_min = q_min.min(point.q);
+                q_max = q_max.max(point.q);
+
+                r_min = r_min.min(point.r);
+                r_max = r_max.max(point.r);
+
+                let s = point.compute_s();
+                s_min = s_min.min(s);
+                s_max = s_max.max(s);
+            }
+
+            Inequality {
+                q_min,
+                q_max,
+                r_min,
+                r_max,
+                s_min,
+                s_max,
+            }
+        }
+
+        pub fn solve<T, D>(&self, data: Option<D>, constuctor: &dyn Fn(Axial, Option<D>) -> T) -> Vec<T>
+        where
+            D: Clone
+        {
+            let mut vec = vec![];
+            for q in self.q_min..=self.q_max {
```

Comment:
Gut feeling that I do not have any backing for, there may be a more idiomatically "rustacean" way to do this.

Reply:
True, I refactored it to not mutate a struct but rather utilize maps and iterators to construct the vector in place.

Reply:
This looks more idiomatic for sure.